### ISSUE TITLE
 Module Federation Support 

### ISSUE THREAD

**Issue by ScriptedAlchemy** [vercel/next.js#16368]
# Feature request

## Is your feature request related to a problem? Please describe.

Support Module Federation ability to share vendor code, like react

## Describe the solution you'd like

We need async boundary before react or react-dom are required

## Describe alternatives you've considered
Below is a partial implementation, but I am having issues with pages being entrypoints, causing react to not resolve as there's no async bound during the bootup
#16176

**Comment by ScriptedAlchemy** [comment]
Update on this. 

I've found a way to move the async boundary into the webpack runtime. This circumvents limitations of next.js by changing how webpack startup runtime module works. 

Replacing the startup module will effectively prevent next from initializing till webpack container negotiations take place. 

I only uncovered this today, so it'll take some time to modify and PR webpack.

**Comment by ckken** [comment]
mark

**Comment by ScriptedAlchemy** [comment]
@rauchg @timneutkens @Timer @sokra I've updated the git issue, please let me know if there's any more information I can provide

**Comment by mlstubblefield** [comment]
Awesome! We're super excited about utilizing next to support microfront ends.

And for what it's worth (seems like everyone has a different definition of micro front ends) here's the short definition of what we're wanting to do.

We want to be to deploy our header/footer (layout) as a next app in vercel as well as 3-4 pages as separate next js apps in vercel.

Ideally they're stitched together and client side routing is well supported (hopefully the nextjs libraries are shared between the apps).

**Comment by ScriptedAlchemy** [comment]
This is more or less what im looking for as well, seamless page routing & component level ownership ☝️

**Comment by ckken** [comment]
mark

**Comment by ckken** [comment]
mark

**Comment by ScriptedAlchemy** [comment]
I've gotten Module Federation and Next routing to work. https://twitter.com/ScriptedAlchemy/status/1402319022331502592?s=20

This works client-side and probably could work server-side with some tweaks. 

You'll need a catchall page and the nextjs-mf plugin. 

Sample PR: https://github.com/module-federation/module-federation-examples/pull/859

Likely ill move to a switch statement so I can take advantage of the `import()` syntax over the low-level API

Also will probably use a suspense implementation and if need be, write some utility which will be added to next-mf if a well-rounded solution looks cumbersome

**Comment by ScriptedAlchemy** [comment]
As of today, I was able to use just the Module Federation plugin with next js. 

It appears that sharing works now, at least when it's in eager mode. 

As of my testing today - nextjs supports federation without any workarounds. 🎉 

It's almost too good to be true

**Comment by Darth-Knoppix** [comment]
I’m having mixed results getting a good setup with Next.js 11 due to the issues @ScriptedAlchemy describes. Does anyone have an example project using Next.js 11 bi-directional CSR?

**Comment by ScriptedAlchemy** [comment]
you have to use the next-mf plugin, which pretty much makes a third webpack build

**Comment by ecwyne** [comment]
🤔 https://www.npmjs.com/package/@module-federation/nextjs-mf

**Comment by jonniedarko** [comment]
@ecwyne https://github.com/module-federation/nextjs-mf is private repo now and 
https://twitter.com/ScriptedAlchemy/status/1418508699891425283

**Comment by ScriptedAlchemy** [comment]
Yep it's no longer a free plugin sorry fam but time is money 💰 

Next keeps breaking our support and keeping the project working for everyone requires non free engineering resources. 

All / many MF works moving forward will likely cater to paid packages as the OSS R&D is becoming too expensive. 

Times be changing. OSS was good while it lasted but funding is just too low

**Comment by besh** [comment]
@ScriptedAlchemy Thanks for your efforts and understand entirely

> v2 is the last open release but doesn't work with newer versions of nextjs

Do you happen to know the last good combination of next and nextjs-mf? Are you saying that all > v11 next is a non-starter? I'm on `next` `11.0.1` and `@module-federation/nextjs-mf` and `2.3.1` and noticing two things:

- the `vendors-node_modules_*` files that get generated in the chucks folder of the remote app , (e.g. `http://localhost:3001/_next/static/chunks/vendors-node_modules_react_index_js.js`) 404 in the consuming app because the consuming app is searching for them on _its_ host (`http://localhost:3000/_next/static/chunks/vendors-node_modules_react_index_js.js`). I patched this locally with a copy script, but clearly, I'm doing something wrong.
- Regardless of my shared settings for `react` and `react-dom`, I'm facing invalid hook calls.

I'm not necessarily seeking help on fixing those issues specifically, more so wondering if I'm barking up the wrong tree by attempting to use next 11.

**Comment by TychosDrunkElk** [comment]
@besh I was able to get next 11 and nextjs-mf 2.3.1 working with a few tweaks (private project, so i cant share but will try to put up an example in a bit):
- make sure you are setting `config.output.publicPath = 'auto'` in your remote app's webpack config
- in the consuming app at the top of your `_app` file, add:
```js
if (process.browser) {
  window.remoteApp.init(__webpack_share_scopes__.default);
}
``` 
(see https://webpack.js.org/concepts/module-federation/#dynamic-remote-containers)
- make sure react is shared in both the consuming and remote app, with the consuming app having `eager: true` in addition to `requiredVersion: false` and `singleton: true`.

Obviously this wont work for ssr though, excited to see progress made in that area!

**Comment by TychosDrunkElk** [comment]
OH also I had to use `window.remoteApp.get` instead of dynamic imports, probably for the same reason described in https://app.privjs.com/buy/packageDetail?pkg=@module-federation/nextjs-mf.

**Comment by besh** [comment]
@TychosDrunkElk nailed it. `config.output.publicPath = 'auto' ` resolved the chunk issue and`  window.remoteApp.init(__webpack_share_scopes__.default);` resolved the sharing issue. I personally did not have to use `window.{the_remove_app}.get` in my case, but will keep an eye out for issues there.

Thank you for sharing.

**Comment by ScriptedAlchemy** [comment]
yeah if you use a federated import in many dependency chains attached to index.js, it breaks the build - you have to use the low-level API

**Comment by lu-zen** [comment]
Looks like Vercel is ignoring all effort you're making for an entire year, what a shame.

**Comment by lu-zen** [comment]
@ScriptedAlchemy did you ever explored if it's possible to use MF with [Remix](https://docs.remix.run/)?

**Comment by ScriptedAlchemy** [comment]
> @ScriptedAlchemy did you ever explored if it's possible to use MF with [Remix](https://docs.remix.run/)?

Yeah it is. Jacob used to be on my platform team at lulu. Hems at remixx now. He knows MF better than anyone else outside of webpack, together we got MF working on next and in the process  he prototyped it on remixx

The trick is you pretty much have a sidecar build (same as next) that's webpack based. So the remote entry and exposed modules go thru webpack. Then you can hook share scope up to ESM modules from esbuild. 

Process is near the same for how we make it work in nextjs. Sometime in December we were planning to try and make something  out of it. 

Spoke with him about it last week actually so this is something we are actively thinking about

**Comment by yumauri** [comment]
So... Is it currently possible to use Next 12 with Module Federation? Couldn't find any repos with examples, only with previous version.

**Comment by victordidenko** [comment]
@ScriptedAlchemy does paid version of `@module-federation/nextjs-mf` supports Next 12 yet?

**Comment by ScriptedAlchemy** [comment]
Last I tested it works just fine. Nothing next can change impacts my plugin. Federated modules are compiled independently of nexts actual build. 

I updated my largest repo today and federation required no change.  Mostly Babel problems on the main build lol

**Comment by ScriptedAlchemy** [comment]
> @ScriptedAlchemy did you ever explored if it's possible to use MF with [Remix](https://docs.remix.run/)?

Yeah we already got it to work months ago. Sometime in December me and Jacob are going to pair up and improve it. 

You can share any module with any system back to webpack. So esbuild and ESM modules can be shared. But only webpack can expose modules. So we do the same thing I did to next. Use webpack only for exposed modules and wire up the apis under the hood

**Comment by ScriptedAlchemy** [comment]
Total speculation. But if vercel plans to support federation specifically it probably will happen when react 18 is released. 

I do have federated SSR working already (for like 6 months now) 

But R18 would simplify it a little and full async support would be prime for MF support

**Comment by sirajtechy** [comment]
This Feature is the game changer and the Next js team still thinking to prioritise it. Development becoming costlier these days and Time is becoming so essential to build out a ROBUST, RESILIENT, RELIABLE product. When @ScriptedAlchemy  like people are shedding time and effort to bring out the crazy feature up and running for the FE community to leverage it why not WE Embrace it. !! On a side note I am not commercializing Zack here for his effort but truly respect his work to solve the daily problems we engineers face during the development and deployment process :-(

**Comment by ScriptedAlchemy** [comment]
Apply enough pressure, and vercel will bend to support it. I imagine it's on the roadmap, but if this git issue got a few hundred comments - it would be harder to ignore. 

So if you want it sooner and don't want to depend on my commercial packages, get as many developers as possible to comment on this issue. 

Strength in numbers and the squeaky wheel gets the grease. We all want it but we need to come together and apply a lot more pressure to make this a top priority item. 

I'm not a fan of grayhat tactics. But if you all want this capability sooner - you need to broadcast and get as many people as possible to join this conversation. 

Vercel has all the resources needed to make this a reality. I've seen what Tobias can do in mere hours. He runs circles around me.  

Chances are he could bring first class support to next within a week if there's enough demand from end users. Gut feeling is this is a product issue and the jira ticket keeps getting pushed down the backlog. 

Just saying... The founder of webpack is on staff at this company, this isn't rocket science especially to the author of the compiler itself

**Comment by lednhatkhanh** [comment]
We really really want to have native module federation support in Next.js!

**Comment by sirajtechy** [comment]
@ScriptedAlchemy 
I am already buying ur library and sooner for my Org project too :-)

**Comment by mehmetpekcan** [comment]
Native module federation support would be great and on the other hand I appreciate Zack for all of his efforts. He is doing great things on this way!

**Comment by marcelovicentegc** [comment]
I have been following this issue for quite some time now and have [tried to figure out ways to build NextJS micro-frontend frameworks without Module Federation](https://github.com/micro-frontend-frameworks). I'm considering jumping into making NextJS and Module Federation compatible, but that would mean a lot of time put in since I have no context of how both work internally as of now, but that seems like the unbeatable way of building MFEs with NextJS. 

Zack, thanks for all your efforts! Vercel folks, thanks for bringing the best React framework to life. NextJS being compatible with Module Federation will be revolutionary.

**Comment by noreiller** [comment]
I have made Next.js work with Module Federation thanks to the work of @ScriptedAlchemy and @telenko with [node-mf](https://github.com/telenko/node-mf).
It supports server and client side but I missed time to contribute back to the community. I hope to be able to do it next week. 🤞

**Comment by BhaskaranR** [comment]
we are using nextjs for backoffice tools. it will be helpful if you could support module federation out of box.

**Comment by alinaMihai** [comment]
I would appreciate it very much if Module Federation would be available with Next.js.  My team and I would have loved to use Module Federation as being the most optimal way to work with Micro Frontends. Currently, we are trying to use Next.js Zones, and there are some drawbacks as you can't really share code between apps.

**Comment by tundera** [comment]
This is an awesome feature and would only make Next.js even better

**Comment by SketchLagoon** [comment]
This would be an invaluable tool in the toolbox for nextjs! 🙌 It's the one thing holding me back from integrating nextjs into some of my projects.

**Comment by nguyenyou** [comment]
Please make it happen! 🙏

**Comment by ScriptedAlchemy** [comment]
> I have made Next.js work with Module Federation thanks to the work of @ScriptedAlchemy and @telenko with [node-mf](https://github.com/telenko/node-mf).
> It supports server and client side but I missed time to contribute back to the community. I hope to be able to do it next week. 🤞

Let’s talk because I’m looking at doing the same thing.

**Comment by noreiller** [comment]
> > I have made Next.js work with Module Federation thanks to the work of @ScriptedAlchemy and @telenko with [node-mf](https://github.com/telenko/node-mf).
> > It supports server and client side but I missed time to contribute back to the community. I hope to be able to do it next week. 🤞
> 
> Let’s talk because I’m looking at doing the same thing.

Yes, I will contact you next week (I'm in vacation right now).

**Comment by rmunch** [comment]
Would love to see native Module Federation support in NextJS as well 👍

> I have made Next.js work with Module Federation thanks to the work of @ScriptedAlchemy and @telenko with [node-mf](https://github.com/telenko/node-mf). It supports server and client side but I missed time to contribute back to the community. I hope to be able to do it next week. 🤞

I've also got client and server federation working thanks to this prior art. It was necessary to add async boundaries in certain places since NextJS doesn't have a central boot point, though I was able to figure out a pattern to do this at the page level and to enforce it with a custom ESLint ruleset - happy to share details if it'd be helpful.

**Comment by ScriptedAlchemy** [comment]
Im glad we can make this work and bridge some gaps. But it’s important to note that us making it “work” is a massive workaround that literally adds 4 webpack builds to the app. 

It’s very important vercel implements support for this feature because we are all building the codebase 4 times to make it work right now

**Comment by acojuli** [comment]
We love to have native module federation support in Nextjs !

**Comment by lu-zen** [comment]
Would like to see this feature natively supported in next, it makes mfe so easy, many teams will benefit from this.

**Comment by danantal** [comment]
This is the kind of feature that would make next be one step ahead over competitors.

**Comment by nodkz** [comment]
Module federation in Next.js will be the next step for [Spotify's BFF architecture](https://philcalcado.com/2015/09/18/the_back_end_for_front_end_pattern_bff.html) when become able to develop frontend UI inside Service teams. It's incredibly important in complex data domains when knowledge distribution between developers is very difficult.

I was expecting the ModuleFederation to appear in NextJS since version 11. And I'm still can't understand why this important feature is not implemented yet.

Right now we have dozens of separate NextJs applications and navigation between them enforces full page reload. And it's a bad user experience for our users but it is less evil according to the complexity of our data domains. 

Engineering consists of trade-offs. And we are building our systems from trade-offs. I hope and believe that someday MF will be available in NextJS out of the box.

**Comment by ScriptedAlchemy** [comment]
Update: I just got Federation working for both client and server via code streaming. I should be ready to release a new plugin in the next week or so. 

Still this is a workaround solution for federation on next, but at least federated code will work beyond the client-side. 
This won't utilize our new federated multiplexing so you need rolling restarts to re-require new federated module updates. But once we have multiplexing finished, one would be able to securely stream any software. Multiplexing will bring us to a state of zero-install, always on, always warm servers that never require a redeploy. 
It should work with next.js as well since the multiplexing plugin will be altering how webpack loads and runs modules, but multiplexing is designed to stream any language from any location to any machine on demand. 

First stop is federated Cloudflare workers, and yes, you'd be able to deploy a standard next.js app to a Cloudflare worker. Not the SSG or Cloudflare Pages half-measure but full next.js apps as they exist today running on the edge with no backing Lambda or Infrastructure whatsoever. 

Handles multiple files, dynamic imports, code streams, and can execute up to 25mb chunks (can execute multiple chunks) 

Initial testing shows "cold start" times of 0.5ms and average uncached execution duration for a SSR app is around 25ms :)

**Comment by Marcelx8** [comment]
We would love to use Module Federation with Nextjs SSR to help with our Micro Frontend approach. Nextjs can definitely take lead in this feature.

**Comment by noreiller** [comment]
[node-mf](https://github.com/telenko/node-mf) is now supporting Next.js for the server-side part.
Now, we are able to consume a getServerSideProps function from another app.

The client-side part (including hydration) needs more work but it's a first step.

**Comment by ericvel** [comment]
I am currently working on a big web application, where multiple teams will develop separate features that will all be part of the same shared container (header, footer etc.). Using module federation to achieve this seems like the natural way to go, and having it natively supported by Next.js would make our lives infintely easier. 

Please don't ignore this issue, Vercel!

**Comment by alexUXUI** [comment]
As an architect at large enterprises, I work with systems at scale which often require distributing code. Module federation perfectly enables the ability to distribute code and I genuinely think Next.js will capture more of the enterprise market with MF support. Let’s make Next.js the react meta framework for enterprise level needs. This could be a massive point of differentiation from solutions like Remix. Thanks a bunch for your consideration.

**Comment by lancej1022** [comment]
> > @ScriptedAlchemy did you ever explored if it's possible to use MF with [Remix](https://docs.remix.run/)?
> 
> Yeah it is. Jacob used to be on my platform team at lulu. Hems at remixx now. He knows MF better than anyone else outside of webpack, together we got MF working on next and in the process he prototyped it on remixx
> 
> The trick is you pretty much have a sidecar build (same as next) that's webpack based. So the remote entry and exposed modules go thru webpack. Then you can hook share scope up to ESM modules from esbuild.
> 
> Process is near the same for how we make it work in nextjs. Sometime in December we were planning to try and make something out of it.
> 
> Spoke with him about it last week actually so this is something we are actively thinking about

@ScriptedAlchemy do you know if there are any plans to add an example of this to the Module Federation examples on GitHub and/or release a plugin to facilitate this?

We already use Module Federation at work but are currently doing client-rendering which has obvious paint points. Ive wanted to move us to Next for some time now, but the ongoing lack of support for Module Federation means we would need to do some drastic reworking of our architecture that just wouldn't be worth it. 

If Remix looks winds up with more realistic Module Federation support in the nearer future, then that's ultimately the path for our company to head down.

**Comment by yumauri** [comment]
> If Remix looks winds up with more realistic Module Federation support in the nearer future, then that's ultimately the path for our company to head down.

@lancej1022 I pretty much doubt Remix will adopt MF in near future. Reason is dead simple — Remix doesn't use Webpack.

**Comment by balazsorban44** [comment]
We appreciate everyone letting us know they’re looking for a solution here. However, we'd like to ask you to stop commenting unless you have additional info to move this discussion forward. If you would like to show your support, please add a :thumbsup: emoji on the initial comment instead. Thank you.
